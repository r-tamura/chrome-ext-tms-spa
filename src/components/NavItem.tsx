import * as React from "react"
import { Link } from "react-router-dom"

interface INavProps extends React.ClassAttributes<{}> {
  to: string,          // Routeのナビゲーション先
  selected?: boolean,  // 選択中であるか
  disabled?: boolean,  // 無効リンクであるか
}

/**
 * ナビゲーションアイテムコンポーネント
 * サイドナビゲーションの項目
 *
 * @param {string} to Routeのナビゲーション先
 * @param {boolean} selected 選択中であるか(true: 選択中/false: 未選択)
 * @param {boolean} disabled 無効なリンクであるか(true: 無効/false: 有効)
 * @param {React.ReactNode} children 子コンポーネント
 */
const NavItem: React.SFC<INavProps> = ({
  to,
  disabled = false,
  selected = false,
  children,
}) => {
  // Classes
  const defaultClass = "menu-item nav-item"
  const classes = [
    defaultClass,
    selected ? "selected" : "",
    disabled ? "disabled" : "",
  ].filter(s => s !== "").join(" ")

  // Attributes
  const attrs = {
    disabled,
  }
  return disabled
      ? <div className={classes} {...attrs}>{children}</div>
      : <Link to={to} className={classes} {...attrs}>{children}</Link>
}

export {
  INavProps,
}

export default NavItem
