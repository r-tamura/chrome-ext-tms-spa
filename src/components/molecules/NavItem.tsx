import React, { useContext } from "react";
import { Link } from "react-router-dom";
import { MenuItem } from "./Menu";
import styled, { ThemeContext } from "styled-components";

interface INavProps {
  to: string; // Routeのナビゲーション先
  selected?: boolean; // 選択中であるか
  disabled?: boolean; // 無効リンクであるか
}

/**
 * ナビゲーションアイテムコンポーネント
 * サイドナビゲーションの項目
 *
 * @param {string} to Routeのナビゲーション先
 * @param {boolean} selected 選択中であるか(true: 選択中/false: 未選択)
 * @param {boolean} disabled 無効なリンクであるか(true: 無効/false: 有効)
 * @param {React.ReactNode} children 子コンポーネント
 */
export const NavItem: React.SFC<INavProps> = ({
  to,
  disabled = false,
  selected = false,
  children
}) => {
  const theme = useContext(ThemeContext);
  const StyledLink = disabled ? DisabeldLink : BaseLink;

  return (
    <MenuItem
      disabled={disabled}
      style={
        selected
          ? {
              backgroundColor: theme.primaryDark
            }
          : null
      }
    >
      <StyledLink to={to}>{children}</StyledLink>
    </MenuItem>
  );
};

const BaseLink = styled(Link)`
  color: ${({ theme }) => theme.white};
  display: block;
`;

const DisabeldLink = styled.div`
  color: ${({ theme }) => theme.white};
  opacity: 0.7;
`;
