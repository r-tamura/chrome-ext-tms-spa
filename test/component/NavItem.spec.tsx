import React from "react";
import renderer from "react-test-renderer";
import { mount, render } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";
import NavItem, { INavProps } from "~/components/NavItem";

const component = (props: INavProps = { to: "/" }) => (
  // keyを与えないとスナップショットテストに失敗する
  // refers: https://github.com/ReactTraining/react-router/issues/5579
  <MemoryRouter initialEntries={[{ pathname: "/", key: "testKey" }]}>
    <NavItem {...props}>Item</NavItem>
  </MemoryRouter>
);

describe("<NavItem />", () => {
  it("snap shot", () => {
    const c = renderer.create(component()).toJSON();
    expect(c).toMatchSnapshot();
  });

  it("full DOM rendering", () => {
    const w = mount(component({ to: "/" }));
    expect(toJson(w)).toMatchSnapshot();
    expect(w.text()).toBe("Item");

    const $link = w.find("Link");
    expect($link.length).toBe(1);
    expect($link.prop("to")).toBe("/");
  });

  it("should be selected", () => {
    const w = render(component({ to: "/", selected: true }));
    expect(w.hasClass("selected")).toBeTruthy();
  });

  it("should be disabled", () => {
    const w = render(component({ to: "/", disabled: true }));
    expect(w.is("div")).toBe(true);
    expect(w.hasClass("disabled")).toBeTruthy();
  });

  it("static rendering", () => {
    const w = render(component({ to: "/" }));
    expect(w.is("a")).toBe(true);
    expect(w.hasClass("selected")).toBeFalsy();
    expect(w.hasClass("disabled")).toBeFalsy();
  });
});
