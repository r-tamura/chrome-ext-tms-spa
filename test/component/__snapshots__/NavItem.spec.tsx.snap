// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavItem /> full DOM rendering 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <NavItem
      to="/"
    >
      <Link
        className="menu-item nav-item"
        disabled={false}
        to="/"
      >
        <a
          className="menu-item nav-item"
          disabled={false}
          href="/"
          onClick={[Function]}
        >
          Item
        </a>
      </Link>
    </NavItem>
  </Router>
</MemoryRouter>
`;

exports[`<NavItem /> snap shot 1`] = `
<a
  className="menu-item nav-item"
  disabled={false}
  href="/"
  onClick={[Function]}
>
  Item
</a>
`;
